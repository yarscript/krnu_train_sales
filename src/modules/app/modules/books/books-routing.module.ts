import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';

import {
  CollectionPageComponent,
  FindBookPageComponent,
  ViewBookPageComponent,
}                          from '@/modules/app/modules/books/containers';
import { BookExistsGuard } from '@/modules/app/modules/books/guards';

export const routes: Routes = [
  {
    path: 'find',
    component: FindBookPageComponent,
    data: { title: 'Find book' },
  },
  {
    path: ':id',
    component: ViewBookPageComponent,
    canActivate: [BookExistsGuard],
    data: { title: 'Book details' },
  },
  {
    path: '',
    component: CollectionPageComponent,
    data: { title: 'Collection' },
  },
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
})
export class BooksRoutingModule {}
