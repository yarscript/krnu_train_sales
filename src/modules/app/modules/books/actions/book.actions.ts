import { createAction, props } from '@ngrx/store';

import { Book } from '@/modules/app/modules/books/models';

export const loadBook = createAction(
  '[Book Exists Guard] Load Book',
  props<{ book: Book }>()
);
