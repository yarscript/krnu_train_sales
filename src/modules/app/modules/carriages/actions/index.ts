import * as CarriageActions           from './carriage.actions';
import * as CarriageApiActions        from './carriage-api.actions';
import * as CarriageCreatePageActions from './create-page.actions';

export { CarriageActions, CarriageApiActions, CarriageCreatePageActions }
