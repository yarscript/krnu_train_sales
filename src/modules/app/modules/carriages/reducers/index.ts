import {
  createSelector, createFeatureSelector, Action, combineReducers,
} from '@ngrx/store';

import * as fromRoot          from '@/modules/app/reducers';
import * as fromCreatePage    from '@/modules/app/modules/carriages/reducers/create-page.reducer';
import * as fromCarriageReducers from '@/modules/app/modules/carriages/reducers/carriages.reducer';


export const organisationFeatureKey = 'carriages';


export interface CarriageState
{
  [fromCarriageReducers.CarriageFeatureKey]: fromCarriageReducers.State;
  [fromCreatePage.createPageFeatureKey]: fromCreatePage.State;
}

export interface State extends fromRoot.State
{
  [organisationFeatureKey]: CarriageState;
}

export function reducers(state: CarriageState | undefined, action: Action) {
  return combineReducers({
    [fromCreatePage.createPageFeatureKey]: fromCreatePage.reducer,
    [fromCarriageReducers.CarriageFeatureKey]: fromCarriageReducers.reducer,
  })(state, action);
}

export const selectCarriagesState = createFeatureSelector<State, CarriageState>(
  organisationFeatureKey
);

export const selectCarriageEntitiesState = createSelector(
  selectCarriagesState, (state) => state.carriage
);
export const selectCarriages = createSelector(
  selectCarriageEntitiesState, fromCarriageReducers.getCarriages
);


export const selectCreatePageState = createSelector(
  selectCarriagesState, (state) => state.createPage
);
export const selectCreatePageError = createSelector(
  selectCreatePageState, fromCreatePage.getError
);
export const selectCreatePagePending = createSelector(
  selectCreatePageState, fromCreatePage.getPending
);
