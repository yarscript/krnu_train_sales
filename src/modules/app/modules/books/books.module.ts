import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';

import { BooksRoutingModule }             from '@/modules/app/modules/books/books-routing.module';
import {
  BookAuthorsComponent,
  BookDetailComponent,
  BookPreviewComponent,
  BookPreviewListComponent,
  BookSearchComponent,
}                                         from '@/modules/app/modules/books/components';
import {
  CollectionPageComponent,
  FindBookPageComponent,
  SelectedBookPageComponent,
  ViewBookPageComponent,
}                                         from '@/modules/app/modules/books/containers';
import { BookEffects, CollectionEffects } from '@/modules/app/modules/books/effects';

import * as fromBooks     from '@/modules/app/modules/books/reducers';
import { MaterialModule } from '@/modules/shared/modules/material';
import { PipesModule }    from '@/modules/shared/pipes';

export const COMPONENTS = [
  BookAuthorsComponent,
  BookDetailComponent,
  BookPreviewComponent,
  BookPreviewListComponent,
  BookSearchComponent,
];

export const CONTAINERS = [
  FindBookPageComponent,
  ViewBookPageComponent,
  SelectedBookPageComponent,
  CollectionPageComponent,
];

@NgModule({
  imports: [
    CommonModule,
    MaterialModule,
    BooksRoutingModule,

    /**
     * StoreModule.forFeature is used for composing state
     * from feature modules. These modules can be loaded
     * eagerly or lazily and will be dynamically added to
     * the existing state.
     */
    StoreModule.forFeature(fromBooks.booksFeatureKey, fromBooks.reducers),

    /**
     * Effects.forFeature is used to register effects
     * from feature modules. Effects can be loaded
     * eagerly or lazily and will be started immediately.
     *
     * All Effects will only be instantiated once regardless of
     * whether they are registered once or multiple times.
     */
    EffectsModule.forFeature([BookEffects, CollectionEffects]),
    PipesModule,
  ],
  declarations: [COMPONENTS, CONTAINERS],
})
export class BooksModule {}
