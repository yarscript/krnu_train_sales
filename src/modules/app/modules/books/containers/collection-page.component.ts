import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';

import { Store } from '@ngrx/store';
import { Observable } from 'rxjs';

import { CollectionPageActions } from '@/modules/app/modules/books/actions';
import { Book }                  from '@/modules/app/modules/books/models';
import * as fromBooks            from '@/modules/app/modules/books/reducers';

@Component({
  selector: 'app-collection-page',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <mat-card>
      <mat-card-title>My Collection</mat-card-title>
    </mat-card>

    <app-book-preview-list [books]="books$ | async"></app-book-preview-list>
  `,
  /**
   * Container components are permitted to have just enough styles
   * to bring the view together. If the number of styles grow,
   * consider breaking them out into presentational
   * components.
   */
  styles: [
    `
      mat-card-title {
        display: flex;
        justify-content: center;
      }
    `,
  ],
})
export class CollectionPageComponent implements OnInit {
  books$: Observable<Book[]>;

  constructor(private store: Store<fromBooks.State>) {
    this.books$ = store.select(fromBooks.selectBookCollection);
  }

  ngOnInit() {
    this.store.dispatch(CollectionPageActions.enter());
  }
}
