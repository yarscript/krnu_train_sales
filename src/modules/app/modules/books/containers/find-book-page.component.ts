import { ChangeDetectionStrategy, Component } from '@angular/core';

import { Store } from '@ngrx/store';
import { Observable } from 'rxjs';
import { take } from 'rxjs/operators';

import { FindBookPageActions } from '@/modules/app/modules/books/actions';
import { Book }                from '@/modules/app/modules/books/models';
import * as fromBooks          from '@/modules/app/modules/books/reducers';

@Component({
  selector: 'app-find-book-page',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <app-book-search
      [query]="searchQuery$ | async"
      [searching]="loading$ | async"
      [error]="error$ | async"
      (search)="search($event)"
    >
    </app-book-search>
    <app-book-preview-list [books]="books$ | async"> </app-book-preview-list>
  `,
})
export class FindBookPageComponent {
  searchQuery$: Observable<string>;
  books$: Observable<Book[]>;
  loading$: Observable<boolean>;
  error$: Observable<string>;

  constructor(private store: Store<fromBooks.State>) {
    this.searchQuery$ = store.select(fromBooks.selectSearchQuery).pipe(take(1));
    this.books$ = store.select(fromBooks.selectSearchResults);
    this.loading$ = store.select(fromBooks.selectSearchLoading);
    this.error$ = store.select(fromBooks.selectSearchError);
  }

  search(query: string) {
    this.store.dispatch(FindBookPageActions.searchBooks({ query }));
  }
}
