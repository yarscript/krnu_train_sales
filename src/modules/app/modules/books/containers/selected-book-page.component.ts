import { ChangeDetectionStrategy, Component } from '@angular/core';

import { Store } from '@ngrx/store';
import { Observable } from 'rxjs';

import { SelectedBookPageActions } from '@/modules/app/modules/books/actions';
import { Book }                    from '@/modules/app/modules/books/models';
import * as fromBooks              from '@/modules/app/modules/books/reducers';

@Component({
  selector: 'app-selected-book-page',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <app-book-detail
      [book]="book$ | async"
      [inCollection]="isSelectedBookInCollection$ | async"
      (add)="addToCollection($event)"
      (remove)="removeFromCollection($event)"
    >
    </app-book-detail>
  `,
})
export class SelectedBookPageComponent {
  book$: Observable<Book>;
  isSelectedBookInCollection$: Observable<boolean>;

  constructor(private store: Store<fromBooks.State>) {
    this.book$ = store.select(fromBooks.selectSelectedBook) as Observable<Book>;
    this.isSelectedBookInCollection$ = store.select(
      fromBooks.isSelectedBookInCollection
    );
  }

  addToCollection(book: Book) {
    this.store.dispatch(SelectedBookPageActions.addBook({ book }));
  }

  removeFromCollection(book: Book) {
    this.store.dispatch(SelectedBookPageActions.removeBook({ book }));
  }
}
