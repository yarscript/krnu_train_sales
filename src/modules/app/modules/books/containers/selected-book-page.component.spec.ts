import { ComponentFixture, TestBed } from '@angular/core/testing';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';

import { MockStore, provideMockStore } from '@ngrx/store/testing';

import { SelectedBookPageActions }   from '@/modules/app/modules/books/actions';
import {
  BookAuthorsComponent,
  BookDetailComponent,
}                                    from '@/modules/app/modules/books/components';
import { SelectedBookPageComponent } from '@/modules/app/modules/books/containers/index';
import { Book, generateMockBook }    from '@/modules/app/modules/books/models';
import * as fromBooks                from '@/modules/app/modules/books/reducers';
import { AddCommasPipe }             from '@/modules/shared/pipes/add-commas.pipe';
import { MaterialModule }            from '@/modules/shared/modules/material';

describe('Selected Book Page', () => {
  let fixture: ComponentFixture<SelectedBookPageComponent>;
  let store: MockStore;
  let instance: SelectedBookPageComponent;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [NoopAnimationsModule, MaterialModule],
      declarations: [
        SelectedBookPageComponent,
        BookDetailComponent,
        BookAuthorsComponent,
        AddCommasPipe,
      ],
      providers: [provideMockStore()],
    });

    fixture = TestBed.createComponent(SelectedBookPageComponent);
    instance = fixture.componentInstance;
    store = TestBed.inject(MockStore);

    spyOn(store, 'dispatch');
  });

  it('should compile', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch a collection.AddBook action when addToCollection is called', () => {
    const $event: Book = generateMockBook();
    const action = SelectedBookPageActions.addBook({ book: $event });

    instance.addToCollection($event);

    expect(store.dispatch).toHaveBeenLastCalledWith(action);
  });

  it('should dispatch a collection.RemoveBook action on removeFromCollection', () => {
    const $event: Book = generateMockBook();
    const action = SelectedBookPageActions.removeBook({ book: $event });

    instance.removeFromCollection($event);

    expect(store.dispatch).toHaveBeenLastCalledWith(action);
  });
});
