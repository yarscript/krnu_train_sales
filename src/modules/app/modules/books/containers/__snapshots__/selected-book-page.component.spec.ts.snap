// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selected Book Page should compile 1`] = `
<app-selected-book-page
  book$={[Function Store]}
  isSelectedBookInCollection$={[Function Store]}
  store={[Function MockStore]}
>
  <app-book-detail>

  </app-book-detail>
</app-selected-book-page>
`;
